### ISSUE TITLE
X4 Foundations: GPU fault

### ISSUE THREAD

**Issue by edmondo** [daniel-schuermann/mesa#39]
No issues without ACO (RADV_PERFTEST=llvm).

With ACO it causes a GPU Fault:

`kernel: amdgpu 0000:08:00.0: GPU fault detected: 147 0x0838c402 for process Main() pid 2294 thread Main() pid 2294
kernel: amdgpu 0000:08:00.0:   VM_CONTEXT1_PROTECTION_FAULT_ADDR   0x0000FD07
kernel: amdgpu 0000:08:00.0:   VM_CONTEXT1_PROTECTION_FAULT_STATUS 0x040C4002
kernel: amdgpu 0000:08:00.0: VM fault (0x02, vmid 2, pasid 32782) at page 64775, read from 'TC3' (0x54433300) (196)
`

Kernel: 5.1.0-gentoo
CPU: AMD Ryzen 5 2600
GPU: AMD RX 470 (AMD RADV/ACO POLARIS10 (LLVM 9.0.0))

[aco.gpulock.dmesg.txt](https://github.com/daniel-schuermann/mesa/files/3356948/aco.gpulock.dmesg.txt)

**Comment by shmerl** [comment]
I had a similar experience - the game works briefly in the main menu, and then GPU resets. 

Kernel: 5.1.15
GPU: Vega 56

I also didn't manage to make Mesa Vulkan overlay work with X4 for some reason. Did anyone try that? I was planning to use it to measure performance, so I used:

```
VK_INSTANCE_LAYERS=VK_LAYER_MESA_overlay
VK_LAYER_MESA_OVERLAY_POSITION=bottom-left
```
But the HUD wasn't coming up.

It works with vkcube.

**Comment by daniel-schuermann** [comment]
Thank you! Trying two things could help a bit as I don't own this game:
1) Would you be able to create a debug build of aco and see if it crashes instead of freezing?
2) Could you create a renderdoc capture with RADV_PERFTEST=LLVM of the main menu a brief moment after it would freeze with aco?

**Comment by varris1** [comment]
Hello,
X4 also had GPU hangs with Vega & LLVM in the past. @hakzsam managed to diagnose what the issue was and added a workaround to mesa.

It may be that ACO falls into the same trap:
http://forum.egosoft.com/viewtopic.php?f=180&t=414524&start=270#p4864925
https://patchwork.freedesktop.org/patch/301291/?series=59824&rev=1

**Comment by daniel-schuermann** [comment]
Thx, but this issue seems related to vertex shaders which are compiled with LLVM (even, if ACO is enabled, it only compiles pixel and compute shaders).

**Comment by varris1** [comment]
Ok. Sorry for the noise then

**Comment by edmondo** [comment]
@daniel-schuermann I compiled debugoptimized and yes, now it asserts with following message:

```
X4: ../mesa-9999/src/amd/compiler/aco_spill.cpp:1370: void aco::{anonymous}::assign_spill_slots(aco::{anonymous}::spill_ctx&, unsigned int): Assertion `ctx.reload_count[pair.first] == ctx.reload_count[pair.second]' failed.
```

I'll try to create a renderdoc capture later without ACO.

EDIT: I cannot create a renderdoc capture for X4 with the launch application menu. Is there another way with steam to do it?

EDIT2: Ah, mabye I have it:
```
RADV_PERFTEST=llvm VK_INSTANCE_LAYERS=VK_LAYER_RENDERDOC_Capture ENABLE_VULKAN_RENDERDOC_CAPTURE=1 %command%
```

EDIT3: here the capture (671 MB):
https://drive.google.com/file/d/11lD1dOpRxoLECo_AImFFWcoOxm5d7q6L/view?usp=sharing

**Comment by pendingchaos** [comment]
The GPU fault should be fixed with 0381766ed37e3cd45af120422c6ddd4bf1cf32a5 and the assertion fixed with 4fa0f218b600f08d4020186422f67dbcf3309cf3.

**Comment by edmondo** [comment]
Thanks, both issues are fixed.
